### PR TITLE
#11766 upgrade mypy

### LINKED ISSUE #11766: upgrade mypy
it's been a minute, let's get up to date

### PR DESCRIPTION
**Pull request by glyph**
## Scope and purpose

Fixes #11766

Add a few words about why this PR is needed and what is its scope.
If the associate ticket(s) fully explain the need you can just refer to it/them.

Add any comments about trade-offs (if any) made in this PR and the reasoning behind them.

Add mentions of things that are not covered here and are planed to be done in separate PRs.

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

### CODE REVIEW

File: src/twisted/web/_flatten.py [hunk 1]
```diff
@@ -272,7 +272,7 @@ def keepGoing(
         dataEscaper: Callable[[Union[bytes, str]], bytes] = dataEscaper,
         renderFactory: Optional[IRenderable] = renderFactory,
         write: Callable[[bytes], object] = write,
-    ) -> Generator[Union[Generator, Deferred[Generator]], None, None]:
+    ) -> Generator[Union[Flattenable, Deferred[Flattenable]], None, None]:
```
Comment (glyph) [inline src/twisted/web/_flatten.py:275]: Kinda weird that this worked before!

File: src/twisted/logger/test/test_util.py [hunk 2]
```diff
@@ -130,4 +109,31 @@ def testObserver(e: LogEvent) -> None:
...
+            root=root,
+            o1=o1,
+            o2=o2,
+            o3=o3,
+            o4=o4,
+            o5=o5,
+            p1=p1,
+            p2=p2,
+            oTest=oTest,
+        )
```
Comment (rockstar) [inline src/twisted/logger/test/test_util.py:138]: Are said "shenanigans" a thing that would prevent us from using f-strings here?

File: src/twisted/web/_flatten.py (hunk 1 above)
Comment (rockstar) [reply src/twisted/web/_flatten.py:275]: I just spent 20 minutes trying to figure out why it worked before, and my best guess now is "mypy bug."

**Review by rockstar — APPROVED** [review]
Thanks for doing this. This kinda stuff often gets overlooked but goes a long way in producing quality over time. I had one small question but, honestly, f-strings v `.format` is six of one and half a dozen of the other; if you don't want to address it, don't even worry about it.

File: src/twisted/logger/test/test_util.py (hunk 2 above)
Comment (glyph) [reply src/twisted/logger/test/test_util.py:138]: Yeah. The shenanigan is that we want to grab this "name" attribute that is not properly predeclared, and mypy (rightly) complains if it has visibility into the code that's being executed in the string.

Comment (glyph) [reply src/twisted/logger/test/test_util.py:138]: (fstrings would obviously improve the visibility even more, which would make it worse.)

**Review by adiroiban — APPROVED** [review]
Thanks Paul for the review and Glyph for the mypy updates.

I am doing a rubber-stamp approval here to get this PR cleared.
It looks like Paul is not a member of Twisted.

@glyph feel free to ask a review from the Twisted team if you think that the review from Paul is not enough.

**Comment by glyph** [comment]
@adiroiban Thanks, much appreciated! We need to make the robot allow members accepting reviews from external contributors, so the robot will be credited with the approval, if someone external does an approval.

**Pull request merged**